File: BalancedBinaryTree.py
# Definition for a  binary tree node
class Node:
    def __init__(self, x, left=None, right=None):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
    def height(self, root):
        if root == None:
            return 0 
        return max(self.height(root.left), self.height(root.right)) + 1
    def isBalanced(self, root):
        if root == None:
            return True
        if abs(self.height(root.left) - self.height(root.right)) <= 1:
            return self.isBalanced(root.left) and self.isBalanced(root.right)
        else:
            return False

File: CPP/Second/balanced_binary_tree.cc
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

/*
  好剪枝
*/

class Solution {
public:
    int height(TreeNode * root) {
        if (! root)
            return 0;
        int left = height(root->left);
        int right = height(root->right);
	// abort caculation, 向上传失败信息
        if (left < 0 || right < 0 || abs(left-right) > 1)
            return -1;
        return max(left, right) + 1;    
    }
    
    bool isBalanced(TreeNode *root) {
        return height(root) >= 0;
    }

    int height2(TreeNode * root) {
        if (! root)
            return 0;
        int left = height(root->left);
	if (left < 0)
	    return -1;
        int right = height(root->right);
	// abort caculation, 向上传失败信息
        if (right < 0 || abs(left-right) > 1)
            return -1;
        return max(left, right) + 1;    
    }
    
    
};



int main(int argc, char *argv[])
{

    return 0;
}
